### PR TITLE
add ISO image unmount action when CD-ROM disks is attached

### PR DESCRIPTION
**Pull request by aiwantaozi**
Issue:
https://github.com/rancher/harvester/issues/67

### CODE REVIEW

File: pkg/api/vm/formatter.go [hunk 1]
```diff
@@ -1,25 +1,93 @@
...
 )
 
 const (
-	startVM   = "startVM"
-	stopVM    = "stopVM"
-	restartVM = "restartVM"
+	startVM      = "startVM"
+	stopVM       = "stopVM"
+	restartVM    = "restartVM"
+	unmountCDRom = "unmountCDRom"
```
Comment (gitlawr) [inline pkg/api/vm/formatter.go]: Maybe `ejectCdrom` is a better name for the action? it sounds more common to me and we have the `eject` Linux command.

File: pkg/api/vm/formatter.go [hunk 2]
```diff
@@ -1,25 +1,93 @@
...
 		resource.AddAction(request, restartVM)
+
+		data := resource.APIObject.Data()
+		vmName := data.String("metadata", "name")
+		vmNamespace := data.String("metadata", "namespace")
+
+		vm, err := vf.vmCache.Get(vmNamespace, vmName)
+		if err == nil {
+			copyObj := vm.DeepCopy()
+			if cdRomExisted := unmountCDRomFromVM(copyObj); cdRomExisted {
```
Comment (gitlawr) [inline pkg/api/vm/formatter.go]: why do we call unmount in a formatter? I think we are gonna add the action here if there's any cdrom?

File: pkg/api/vm/formatter.go [hunk 3]
```diff
@@ -1,25 +1,93 @@
...
+func unmountCDRomFromVM(vm *v1alpha3.VirtualMachine) bool {
+	var cdRomExisted bool
+	var dvName, volumeName string
+	if !vmReady.IsTrue(vm) {
+		return cdRomExisted
+	}
+
+	var dvs []cdiv1.DataVolume
+	for _, v := range vm.Spec.DataVolumeTemplates {
+		if v.Spec.Source.HTTP != nil && strings.HasSuffix(v.Spec.Source.HTTP.URL, "iso") {
```
Comment (gitlawr) [inline pkg/api/vm/formatter.go]: This is not the proper way to find a cdrom, we should check the cdrom field: https://kubevirt.io/user-guide/#/creation/disks-and-volumes?id=cdrom

**Comment by aiwantaozi** [comment]
updated

File: pkg/api/vm/formatter.go [hunk 4]
```diff
@@ -19,7 +23,29 @@ func formatter(request *types.APIRequest, resource *types.RawResource) {
 	if resource.APIObject.Data().Bool("spec", "running") {
 		resource.AddAction(request, stopVM)
 		resource.AddAction(request, restartVM)
+
+		var vm v1alpha3.VirtualMachine
+		err := convert.ToObj(resource.APIObject.Data(), &vm)
+		if err == nil {
+			if canEjectCdRom(&vm) {
```
Comment (gitlawr) [inline pkg/api/vm/formatter.go]: We can use a single if

File: pkg/api/vm/handler.go [hunk 5]
```diff
@@ -69,3 +93,59 @@ func (h *vmActionHandler) restartVM(name, namespace string) error {
...
+		return err
+	}
+
+	if !reflect.DeepEqual(vm, vmCopy) {
+		if _, err := h.vms.Update(vmCopy); err != nil {
+			return err
+		}
+	}
+
+	return h.vmis.Delete(namespace, name, &metav1.DeleteOptions{})
```
Comment (gitlawr) [inline pkg/api/vm/handler.go:114]: Is this expected? The vm restarts and services are down when users try to do a cdrom ejection

File: pkg/api/vm/formatter.go (hunk 4 above)
Comment (aiwantaozi) [reply pkg/api/vm/formatter.go]: updated

File: pkg/api/vm/handler.go (hunk 5 above)
Comment (aiwantaozi) [reply pkg/api/vm/handler.go:114]: CD-ROM  ejection takes effect after restarted, we can warn user this action will restart VM when user use this function

**Review by gitlawr — APPROVED** [review]
LGTM

File: pkg/api/vm/formatter.go [hunk 6]
```diff
@@ -3,12 +3,16 @@ package vm
...
 )
 
 const (
-	startVM   = "startVM"
-	stopVM    = "stopVM"
-	restartVM = "restartVM"
+	startVM    = "startVM"
+	stopVM     = "stopVM"
+	restartVM  = "restartVM"
+	ejectCdRom = "ejectCdRom"
```
Comment (guangbochen) [inline pkg/api/vm/formatter.go:15]: let's use `CDRom` instead

File: pkg/api/vm/schema.go [hunk 7]
```diff
@@ -15,22 +15,31 @@ const (
 )
 
 func RegisterSchema(scaled *config.Scaled, server *server.Server) {
+	server.BaseSchemas.MustImportAndCustomize(EjectCdRomActionInput{}, nil)
```
Comment (guangbochen) [inline pkg/api/vm/schema.go:21]: can we please add a comment to describe how the base scheme and resource action is integrated, e.g,

```
// base scheme is used to customize the input data of resource actions, the action input name should start with lower case.
```

**Review by guangbochen — APPROVED** [review]

**Pull request merged**